[DOCS] Creates a data frame examples page (#389)
elastic/stack-docs: docs/en/stack/data-frames/dataframe-examples.asciidoc

[role="xpack"]
[testenv="basic"]
[[dataframe-examples]]
== {dataframe-cap} examples
++++
<titleabbrev>Examples</titleabbrev>
++++

beta[]

This page provides examples of how to use {dataframe-transforms} to derive useful 
insights from your data. For a more detailed, step-by-step example, see 
<<ecommerce-dataframes,Transforming your data with {dataframes}>>.

* <<example-clientips>>

[float]
[[example-clientips]]
=== Finding suspicious client IPs by using scripted metrics

With {dataframe-transforms}, you can use 
{ref}/search-aggregations-metrics-scripted-metric-aggregation.html[scripted 
metric aggregations] on your data. These aggregations are flexible and make 
it possible to perform very complex processing. Let's use scripted metrics to 
identify suspicious client IPs in the web log 
{kibana-ref}/add-sample-data.html[{kib} sample dataset].

> The example below transforms web log data into an entity-centric index where the 

Reviewer: In my opinion, this introductory sentence doesn't give me too much context of what I'm about to see in the example. If we can summarize what the transform is going to do, I think that's more helpful.  Something like:

For example, let's transform the data such that the new index contains the sum of bytes, the number of distinct URLs, and the number of distinct geographic destinations for each client IP. We'll also use a scripted field to count the specific types of HTTP responses that each client IP receives.

Author: You're right. That's why I added the "Based on..." phrases after the questions at the bottom of the doc, but it is a much better solution. I've tried to keep the example relatively short in terms of prose because the API request and the response are fairly long. However, my intention of keeping it short would hurt the overall lucidity of the example. I amended accordingly.

to count specific types of HTTP responses (error, success, and other).
<4> This `bucket_script` calculates the duration of the `clientip` access based 
on the results of the aggregation.

In this example, the API returns the following response (note that this example 
contains the response partially and shows only the first document object):

[source,js]
----------------------------------
{
  "preview" : [
    {
      "geo" : {
        "src_dc" : 12.0,
        "dest_dc" : 9.0
      },
      "clientip" : "0.72.176.46",
      "agent_dc" : 3.0,
      "responses" : {
        "total" : 14.0,
        "counts" : {
          "other" : 0,
          "success" : 14,
          "error" : 0
        }
      },
      "bytes_sum" : 74808.0,
      "timestamp" : {
        "duration_ms" : 4.919943239E9,
        "min" : "2019-06-17T07:51:57.333Z",
        "max" : "2019-08-13T06:31:00.572Z"
      },
      "url_dc" : 11.0
    },
    ...
  }
----------------------------------  
// CONSOLE
// TEST[skip:SETUP]

This {dataframe-transform} makes it easier to answer questions such as:

* Which client IPs are transferring the most amounts of data?

>   (Based on the `"bytes_sum": { "sum": { "field": "bytes" }},` aggregation.)

Reviewer: I think if we set it up properly in the introduction, we don't need to have the "Based on the ..." phrases here.

Author: As above.

[role="xpack"]
[testenv="basic"]
[[dataframe-examples]]
== {dataframe-cap} examples
++++
<titleabbrev>Examples</titleabbrev>
++++

beta[]

> This page provides examples of how to use {dataframe-transforms} to derive useful 

Reviewer: ```suggestion
These examples demonstrate how to use {dataframe-transforms} to derive useful 
```

insights from your data. All the examples use one of the 
{kibana-ref}/add-sample-data.html[{kib} sample datasets]. For a more detailed, 
step-by-step example, see 
<<ecommerce-dataframes,Transforming your data with {dataframes}>>.

* <<example-best-customers>>
* <<example-airline>>
* <<example-clientips>>

[float]
[[example-best-customers]]
=== Finding your best customers

In this example, we use the eCommerce orders sample dataset to find the customers 
who spent the most in our hypothetical webshop. Let's transform the data such 

> that the destination index contains the number of orders, the total price of 

Reviewer: I know I suggested listing these here, but if you find it's too wordy, maybe a solution would be to have a subset, e.g.:

> Let's create/preview?/consider?/look at? a data frame that contains information such as the ..." (and only list a few interesting items).

Author: Maybe it is a bit longer this way, but I think it is also more exact. I would leave as is.

the orders, the amount of unique products and the average price per order, 
and the total amount of ordered products for each customer.

[source,js]
----------------------------------
POST _data_frame/transforms/_preview
{
  "source": {
    "index": "kibana_sample_data_ecommerce"
  },

>   "dest" : {

Reviewer 2: Please add `dest` as a `<numeric-ref>` for all examples - this is the destination index for the data frame and is ignored by `_preview`.

[source,js]
----------------------------------
{
  "preview" : [
    {
      "total_order_amt" : 3946.9765625,
      "order_count" : 59.0,
      "total_unique_products" : 116.0,
      "avg_unique_products_per_order" : 2.0,
      "customer_id" : "10",
      "user" : "recip",
      "avg_amt_per_order" : 66.89790783898304
    },
    ...
    ]
  }
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

This {dataframe} makes it easier to answer questions such as:

* Which customers spend the most?

* Which customers spend the most per order?

* Which customers order most often?

* Which customers ordered the least number of different products?

It's possible to answer these questions using aggregations alone, however 
{dataframes} allow us to persist this data as a customer centric index. This 
enables us to analyze data at scale and gives more flexibility to explore and 
navigate data from a customer centric perspective. In some cases, it can even 
make creating visualizations much simpler.

[float]
[[example-airline]]
=== Finding air carriers with the most delays

In this example, we use the Flights sample dataset to find out which air carrier 

> had the most delays.First, we filter the source data such that it excludes all 

Reviewer: ```suggestion
had the most delays. First, we filter the source data such that it excludes all 
```

          { "term":  { "Cancelled": false } }
        ]
      }
    }
  },
  "dest" : {
    "index" : "sample_flight_delays_by_carrier"
  },
  "pivot": {
    "group_by": { <2>
      "carrier": { "terms": { "field": "Carrier" }}
    },
    "aggregations": {
      "flights_count": { "value_count": { "field": "FlightNum" }},
      "delay_mins_total": { "sum": { "field": "FlightDelayMin" }},
      "flight_mins_total": { "sum": { "field": "FlightTimeMin" }},
      "delay_time_percentage": { <3>
        "bucket_script": {
          "buckets_path": {
            "delay_time": "delay_mins_total.value",
            "flight_time": "flight_mins_total.value"
          },
          "script": "(params.delay_time / params.flight_time) * 100"
        }
      }
    }
  }
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

<1> Filter the source data to select only flights that were not cancelled.
<2> The data is grouped by the `Carrier` field which contains the airline name.
<3> This `bucket_script` performs calculations on the results that are returned 
by the aggregation. In this particular example, it calculates what percentage of 
travel time was taken up by delays.

> The preview {dataframe-transforms} API shows you that the new index would contain 

Reviewer: Not a necessity, but it seems to me you can drop part of this:
```suggestion
The preview shows you that the new index would contain 
```

              state.responses.error += 1 ;
            } else if(code.startsWith('2')) {
              state.responses.success += 1;
            } else {
              state.responses.other += 1;
            }
            """,
          "combine_script": "state.responses",
          "reduce_script": """
            def counts = ['error': 0L, 'success': 0L, 'other': 0L];
            for (responses in states) {
              counts.error += responses['error'];
              counts.success += responses['success'];
              counts.other += responses['other'];
            }
            return counts;
            """
          }
        },
      "timestamp.min": { "min": { "field": "timestamp" }},
      "timestamp.max": { "max": { "field": "timestamp" }},
      "timestamp.duration_ms": { <4>
        "bucket_script": {
          "buckets_path": {
            "min_time": "timestamp.min.value",
            "max_time": "timestamp.max.value"
          },
          "script": "(params.max_time - params.min_time)"
        }
      }
    }
  }
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

<1> This range query limits the transform to documents that are within the 

> last 30 days at the point in time the {dataframe-transform} is started.

Reviewer 2: The range query is evaluated as each checkpoint is processed. For batch data frames, there is only ever 1 checkpoint so this statement is true. However I think we should make this statement also cover continuous data frames which are coming soon. 

Perhaps something along the lines of...
`This range query limits the transform to documents that are within the 
last 30 days at the point in time the {dataframe-transform} checkpoint is processed. For batch data frames this occurs once.`

<2> The data is grouped by the `clientip` field. 
<3> This `scripted_metric` performs a distributed operation on the web log data 
to count specific types of HTTP responses (error, success, and other).
<4> This `bucket_script` calculates the duration of the `clientip` access based 
on the results of the aggregation.

> The preview {dataframe-transforms} API shows you that the new index would contain 

Reviewer: Ditto:
```suggestion
The preview shows you that the new index would contain 
```

POST _data_frame/transforms/_preview
{
  "source": {
    "index": "kibana_sample_data_ecommerce"
  },
  "dest" : {
    "index" : "sample_ecommerce_orders_by_customer"
  },
  "pivot": {
    "group_by": { <1>
      "user": { "terms": { "field": "user" }}, 
      "customer_id": { "terms": { "field": "customer_id" }}
    },
    "aggregations": {
      "order_count": { "value_count": { "field": "order_id" }},
      "total_order_amt": { "sum": { "field": "taxful_total_price" }},
      "avg_amt_per_order": { "avg": { "field": "taxful_total_price" }},
      "avg_unique_products_per_order": { "avg": { "field": "total_unique_products" }},
      "total_unique_products": { "cardinality": { "field": "products.product_id" }}
    }
  }
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

<1> Two `group_by` fields have been selected. This means the {dataframe} will 
contain a unique row per `user` and `customer_id` combination. Within this 
dataset both these fields are unique. By including both in the {dataframe} it 
gives more context to the final results.

NOTE: In the example above, condensed JSON formatting has been used for easier 
readability of the pivot object.

> The API returns the following response (note that this example contains the 

Reviewer: When I first read this I missed that the API was the preview DF API instead of the create DF API. It might be nice to call attention to this and note its usefulness for situations like these where you're playing with DF options. For example, something like this:
 
```suggestion
The preview {dataframe-transforms} API enables you to see the layout of the {dataframe} in advance, populated with some sample values. For example:
```

> response partially and shows only the first document object):

Reviewer: I don't think it's necessary to explain that this is just a subset of the results, since `...` is used pretty widely in our examples.

[source,js]
----------------------------------
{
  "preview" : [
    {
      "total_unique_products" : {
        "avg" : 2.0
      },
      "customer_id" : "10",
      "user" : "recip",

>       "order_id" : {

Reviewer: I think this is the wrong output, since the preview API listed above doesn't include order_id or taxful_total_price, etc. Here's what I get when I run the API on 7.2:

> {
  "preview" : [
    {
      "total_order_amt" : 3946.9765625,
      "order_count" : 59.0,
      "total_unique_products" : 116.0,
      "avg_unique_products_per_order" : 2.0,
      "customer_id" : "10",
      "user" : "recip",
      "avg_amt_per_order" : 66.89790783898304
    },
    ... 
]
}

Reviewer: P.S. I love that these examples use the Kibana sample data, which makes it super easy for me (and our users) to test!

Author: :o Huge thanks for catching this. I played with the API and eventually copied the wrong output.
Lesson learned!

        "value_count" : 59.0
      },
      "taxful_total_price" : {
        "avg" : 66.89790783898304,
        "sum" : 3946.9765625
      },
      "products" : {
        "product_id" : {
          "cardinality" : 116.0
        }
      }
    },
    ...
  }
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

> In the above example we saw how to transform order data into a customer centric 

Reviewer: This "In the above..." sentence is mostly a repeat of what appears in a subsequent paragraph, so I'd remove it here.

> index. Doing this makes it easier to answer questions such as:

Reviewer: ```suggestion
This {dataframe} makes it easier to answer questions such as:
```

* Which customers spend the most?

* Which customers spend the most per order?

* Which customers order most often?

* Which customers ordered the least number of different products?

It's possible to answer these questions using aggregations alone, however 
{dataframes} allow us to persist this data as a customer centric index. This 
enables us to analyze data at scale and gives more flexibility to explore and 
navigate data from a customer centric perspective. In some cases, it can even 
make creating visualizations much simpler.

[float]
[[example-airline]]
=== Finding air carriers with the most delays

> We use the Flights sample dataset to find out which air carrier delayed the most.

Reviewer: I suggest that we introduce this example the same way as we did the first one:
```suggestion
In this example, we use the Flights sample dataset to find out which air carrier delayed the most.
```

> First, we filter the source data such that excludes all the cancelled flights by 

Reviewer: ```suggestion
First, we filter the source data such that it excludes all the cancelled flights by 
```

> using a query filter, then transform the data to contain the distinct number of 

Reviewer: ```suggestion
using a query filter. Then we transform the data to contain the distinct number of 
```

[source,js]
----------------------------------
POST _data_frame/transforms/_preview
{
  "source": {
    "index": "kibana_sample_data_flights",
    "query": { <1>
      "bool": {
        "filter": [
          { "term":  { "Cancelled": false } }
        ]
      }
    }
  },
  "dest" : {
    "index" : "sample_flight_delays_by_carrier"
  },
  "pivot": {
    "group_by": { <2>
      "carrier": { "terms": { "field": "Carrier" }}
    },
    "aggregations": {
      "flights_count": { "value_count": { "field": "FlightNum" }},
      "delay_mins_total": { "sum": { "field": "FlightDelayMin" }},
      "flight_mins_total": { "sum": { "field": "FlightTimeMin" }},
      "delay_time_percentage": { <3>
        "bucket_script": {
          "buckets_path": {
            "delay_time": "delay_mins_total.value",
            "flight_time": "flight_mins_total.value"
          },
          "script": "(params.delay_time / params.flight_time) * 100"
        }
      }
    }
  }
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

<1> Filter the source data to select only flights that were not cancelled.
<2> The data is grouped by the `Carrier` field which contains the airline name.

> <3> This `bucket_script` performs calculations on the results that returned by 

Reviewer: ```suggestion
<3> This `bucket_script` performs calculations on the results that are returned by 
```

> <3> This `bucket_script` performs calculations on the results that returned by 

Reviewer: I think since we link to the scripted_metric reference material in the third example, we should link to https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-pipeline-bucket-script-aggregation.html here too.

Author: I added the link to the first paragraph of the example as I did the same in the third one.

> the aggregation, in this particular example to calculate what percentage of 

Reviewer: ```suggestion
the aggregation. In this particular example, it calculates what percentage of 
```

travel time was taken up by delays.

> The API returns the following response:

Reviewer: I think we can alter this sentence in similar ways to my suggestion for the first example.  Something along the lines of:

> The {dataframe} preview shows you that the new index would contain data like this for each carrier:

[source,js]
----------------------------------
{
  "preview" : [
    {
      "carrier" : "ES-Air",
      "flights_count" : 2802.0,
      "flight_mins_total" : 1436927.5130677223,
      "delay_time_percentage" : 9.335543983955839,
      "delay_mins_total" : 134145.0
    },

>     {

Reviewer: I think you can just show one carrier followed by ..., same as in the first example.

      "carrier" : "JetBeats",
      "flights_count" : 2833.0,
      "flight_mins_total" : 1451143.6898144484,
      "delay_time_percentage" : 8.937088787987832,
      "delay_mins_total" : 129690.0
    },
    {
      "carrier" : "Kibana Airlines",
      "flights_count" : 2832.0,
      "flight_mins_total" : 1419081.404241085,
      "delay_time_percentage" : 9.088273556017194,
      "delay_mins_total" : 128970.0
    },
    {
      "carrier" : "Logstash Airways",
      "flights_count" : 2914.0,
      "flight_mins_total" : 1503620.8713908195,
      "delay_time_percentage" : 9.544959286661593,
      "delay_mins_total" : 143520.0
    }
  ]
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

> The example above transforms flight data into a entity centric index for 

Reviewer: Unless this first sentence would be meaningful to data scientists, I'd omit it since in my opinion it's unnecessary.

> flight carriers. This makes it easier to answer questions such as:

Reviewer: ```suggestion
This {dataframe} makes it easier to answer questions such as:
```

* Which air carrier has the most delays as a percentage of flight time?

> NOTE: Please note that this data is fictional and does not reflect actual delays 

Reviewer: "Please note" is unnecessary inside a "NOTE:"

```suggestion
NOTE: This data is fictional and does not reflect actual delays 
```

or flight stats for any of the featured destination or origin airports.

[float]
[[example-clientips]]
=== Finding suspicious client IPs by using scripted metrics

With {dataframe-transforms}, you can use 
{ref}/search-aggregations-metrics-scripted-metric-aggregation.html[scripted 
metric aggregations] on your data. These aggregations are flexible and make 
it possible to perform very complex processing. Let's use scripted metrics to 
identify suspicious client IPs in the web log sample dataset.

We transform the data such that the new index contains the sum of bytes and the 
number of distinct URLs, agents, incoming requests by location, and geographic 
destinations for each client IP. We also use a scripted field to count the 
specific types of HTTP responses that each client IP receives. Ultimately, the 

> example below transforms web log data into an entity-centric index where the 

Reviewer: Here "entity-centric" has a hyphen and in other cases on this page it does not.  We should be consistent one way or the other.

              counts.error += responses['error'];
              counts.success += responses['success'];
              counts.other += responses['other'];
            }
            return counts;
            """
          }
        },
      "timestamp.min": { "min": { "field": "timestamp" }},
      "timestamp.max": { "max": { "field": "timestamp" }},
      "timestamp.duration_ms": { <4>
        "bucket_script": {
          "buckets_path": {
            "min_time": "timestamp.min.value",
            "max_time": "timestamp.max.value"
          },
          "script": "(params.max_time - params.min_time)"
        }
      }
    }
  }
}
----------------------------------
// CONSOLE
// TEST[skip:SETUP]

<1> This range query limits the transform to documents that are within the 
last 30 days at the point in time the {dataframe-transform} is started.
<2> The data is grouped by the `clientip` field. 
<3> This `scripted_metric` performs a distributed operation on the web log data 
to count specific types of HTTP responses (error, success, and other).
<4> This `bucket_script` calculates the duration of the `clientip` access based 
on the results of the aggregation.

> The API returns the following response (note that this example contains the 

Reviewer: Per suggestions for  first two examples, I recommend changing this introductory sentence to something like this:

> The {dataframe} preview shows you that the new index would contain data like this for each client IP:

response partially and shows only the first document object):

[source,js]
----------------------------------
{
  "preview" : [
    {
      "geo" : {
        "src_dc" : 12.0,
        "dest_dc" : 9.0
      },
      "clientip" : "0.72.176.46",
      "agent_dc" : 3.0,
      "responses" : {
        "total" : 14.0,
        "counts" : {
          "other" : 0,
          "success" : 14,
          "error" : 0
        }
      },
      "bytes_sum" : 74808.0,
      "timestamp" : {
        "duration_ms" : 4.919943239E9,
        "min" : "2019-06-17T07:51:57.333Z",
        "max" : "2019-08-13T06:31:00.572Z"
      },
      "url_dc" : 11.0
    },
    ...
  }
----------------------------------  
// CONSOLE
// TEST[skip:SETUP]

> This {dataframe-transform} makes it easier to answer questions such as:

Reviewer: >  This {dataframe-transform} ....

I used `{dataframe}` in my earlier suggestions (since I thought the transform is was just the job that generates a data frame) but if I've got that wrong please just use the same term in all three places.
